ste/concurrency.go

```diff
@@ -0,0 +1,151 @@
+	// limit of 10,000 OS threads, and panics and shuts down.  This has been observed
+	// on Windows when this value was set to 500 but there were 1000 to 2000 goroutines in the
+	// main pool size.  Using DialContext appears to mitigate that issue, so the value
+	// we compute here is really just to reduce unneeded make and break of connections)
+	s.MaxIdleConnections = maxMainPoolSize
+
+	return s
+}
+
+func getMainPoolSize() (n int, isFromEnv bool) {
+	concurrencyValueOverride := os.Getenv("AZCOPY_CONCURRENCY_VALUE")
+	if concurrencyValueOverride != "" {
+		val, err := strconv.ParseInt(concurrencyValueOverride, 10, 64)
+		if err != nil {
+			log.Fatalf("error parsing the env AZCOPY_CONCURRENCY_VALUE %q failed with error %v",
+				concurrencyValueOverride, err)
+		}
+		return int(val), true
+	}
+
+	numOfCPUs := runtime.NumCPU()
+
+	// fix the concurrency value for smaller machines
+	if numOfCPUs <= 4 {
+		return 32, false
+	}
+
+	// for machines that are extremely powerful, fix to 300 (previously this was to avoid running out of file descriptors, but we have another solution to that now)
+	if 16*numOfCPUs > 300 {
+		return 300, false
+	}
+
+	// for moderately powerful machines, compute a reasonable number
+	return 16 * numOfCPUs, false
+}
+
+// getMaxOpenFiles finds a number of concurrently-openable files
+// such that we'll have enough handles left, after using some as network handles.
+// This is important on Unix, where total handles can be constrained.
+func getMaxOpenPayloadFiles(maxFileAndSocketHandles int, concurrentConnections int) int {
+
+	// The value we return from this routine here only governs payload files. It does not govern plan
+	// files that azcopy opens as part of its own operations.  So we make a reasonable allowance for
+	// how many of those may be opened
+	const fileHandleAllowanceForPlanFiles = 300 // 300 plan files = 300 * common.NumOfFilesPerDispatchJobPart = 3million in total
+
+	const httpHandleAllowanceForOnGoingEnumeration = 1 // might still be scanning while we are transferring. Make this bigger if we ever do parallel scanning
+
+	// make a conservative estimate of total network and file handles known so far
+	estimateOfKnownHandles := int(float32(concurrentConnections)*1.1) +
+		fileHandleAllowanceForPlanFiles +
+		httpHandleAllowanceForOnGoingEnumeration
+
+	// see what we've got left over for open files
+	concurrentFilesLimit := maxFileAndSocketHandles - estimateOfKnownHandles
+
+	// If we get a negative or ridiculously low value, bring it up to some kind of sensible floor
+	// (and take our chances of running out of total handles - which is effectively a bet that
+	// we were too conservative earlier)
+	if concurrentFilesLimit < defaultTransferInitiationPoolSize {
+		concurrentFilesLimit = defaultTransferInitiationPoolSize
+	}
+	return concurrentFilesLimit
+
+	// TODO: consider adding environment var to optionally allow bringing concurrentFilesLimit down lower than what we calculate here
```

Comment:
maybe the user request on limiting simultaneous file uploads can be added

Reply:
@zezha-msft I've added this to the PR. It was more code change than I expected, in order to avoid duplication.  But good to have that setting configurable.  Interested to know if you can think of a better name for the environment variable, and if you think its OK that it doesn't _really_ configure the number of files _exactly_.

It doesn't really configure number of files _exactly_ because number of files != size of transfer initiation pool, in cases where the files are very small and we get unprocessed files queueing up in our chunkfunc channels.  But naming it something like "transfer initiation pool size" in the environment variable would be confusing; and making it _really_ control the number of files (accurately in all cases) would be a bigger code change for little if any benefit.

Reply:
Just waiting on your re-review @zezha-msft , for the changes added in response to this comment